### DISCUSSION TITLE
Uploading to Sharepoint using .NET core

### DISCUSSION THREAD

**Discussion by aathar** [pnp/pnpcore#819] [Q&A] [answered]
Hello

Been trying to upload a file to sharepoint from a web API.
I am able to upload the file using IFormFile, but then the file is uploaded to the server (IIS) first, and then transported to Sharepoint. Any way , I can upload it directly from the api ?

```
 [HttpPost]
        [RequestFormLimits(MultipartBodyLengthLimit = 20971520000)]
        [RequestSizeLimit(20971520000)]
        public async Task<IActionResult> PostFile(IFormFile file)
        {

            using (var pnpContext = await _pnpContextFactory.CreateAsync("Default"))
            {

                // Get a reference to a folder
                IFolder siteAssetsFolder = await pnpContext.Web.Folders.Where(f => f.Name == "Documents").FirstOrDefaultAsync();


                // Upload a file by adding it to the folder's files collection
                IFile addedFile = await siteAssetsFolder.Files
                    .AddAsync(file.Name,
                    System.IO.File.OpenRead($".{Path.DirectorySeparatorChar}tempFiles{Path.DirectorySeparatorChar}{file.Name}"));
}
}
```

I was using pnpCore because I believed that it would be easy to upload a large file using this framework directly to the sharepoint. Can someone please help ?

**Comment by jansenbe** [comment, marked as answer]
@aathar : not sure what the ask is, seems like you want to upload the file directly from your client application without passing via your web API? If so you can upload the file directly from your client side app, assuming you do have a valid authenticated context. Can you provide more details around the ask?

**Reply by aathar** [reply]
hi @jansenbe . Yes thats exactly the ask. I want to upload it directly from application to sharepoint. Here is the scenario:

Say I want to upload a 2 GB file to Sharepoint
User goes on my public website, 
Click on Upload File -> Submit
The file uploads to .NET core
Once the upload is completed, pnpCore code is called to upload it to sharepoint

See, the double file upload is taking place? One to the .NET core application and one to Sharepoint.

I am trying to understand if there is another way of doing this, if I do not want to use the client side app.

Thanks

**Reply by jansenbe** [reply]
If you want stick to server side code you indeed first need to upload the file from your web site to the backend and then the back end uploads it again to SharePoint. Doing this in one go is possible when your web site user authenticates to Azure AD and you then use client side API calls (either REST or Graph) to directly upload the file to SharePoint.
